### PR TITLE
Backup: safe compressor/decompressor closure

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

As it happens, 3rd party compressor/decompressor implementations can be unreliable and hang while `Close()`ing. The current codebase assumes `Close()` always returns in a timely fashion, and therefore the entire backup/restore process can hang while the compressor/decompressor hangs.

There's another issue where if the compressor/decompressor errors, the backup process error flow loses information, as the `Close()` error overrides any prior (and more important) error.

In this PR we do the following:

- Set a timeout for the compressor/decompressor `Close()` function (essentially wrapping up with a `context.WithTimeout()`)
- Ensure to keep track of, and aggregate, all errors up to the `Close()` call.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/13677

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (GuptaManan100) [inline go/vt/mysqlctl/builtinbackupengine.go:859]: You can return the error directly, you don't have to assign to finallErr, that happens implicitly.

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (GuptaManan100) [inline go/vt/mysqlctl/builtinbackupengine.go:887]: Some comment as above and following such assignments and returns. Although I don't mind keeping it this way for readability.

**Review by GuptaManan100 — APPROVED** [review]

**Comment by GuptaManan100** [comment]
Could we add a test to verify the close method returns after 3 seconds, if the decompressor gets stuck?

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go:859]: OMG TIL

Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: OK fixed, basically reverting much into the original assignment logic.

**Comment by shlomi-noach** [comment]
Added unit test for `closeWithTimeout()` function.

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: fixed

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (maxenglander) [inline go/vt/mysqlctl/builtinbackupengine.go:843]: Is the reason for moving this just to be more consistent with where this code is handled in `restoreFile`. I am all for this change just wanted to understand the reasoning.

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (maxenglander) [inline go/vt/mysqlctl/builtinbackupengine.go:765]: Just an idea, we have a package called `ioutil` where we have some utils for customizing reader/writer behavior https://github.com/vitessio/vitess/tree/main/go/ioutil.

You could wrap up this logic in an `ioutil` helper like this:

```
package ioutil

type timeoutCloser struct {
    io.Closer
    ctx context.Context
    timeout time.Duration
}

func (tc *timeoutCloser) Close() {
  // The logic you have below.
}
```

**Review by maxenglander — APPROVED** [review]
A question and suggestion, but LGTM

**Review by deepthi — CHANGES_REQUESTED** [review]
It turns out that in the past, a similar problem was fixed somewhat differently for xtrabackup.
It is worth reading the commentary in #5193 around the change. Once that is done, we'll need to agree on a way to solve it uniformly for both builtin and xtrabackup.
The main points are:
- how long should the timeout be? 3 seconds or 10 minutes?
- is it safe to abandon the Close? #5193 says no, this PR is doing that.

A minor nit regarding go/vt/mysqlctl/builtinbackupengine2_test.go. The filename should be descriptive of what it is intended for. Is there a good reason these tests could not go into the existing test file?

**Comment by shlomi-noach** [comment]
> Is there a good reason these tests could not go into the existing test file?

Yes. For some reason the existing test file uses a different `package` name, and so it's not in the same namespace and we cannot test private functions.

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go:843]: The reason for moving it is to be more `golang` idiomatic: we create the `compressor` in the `if backupStorageCompress` block, and so this is where we should create the deferred `Close()` function, rather than two blocks down the road.

File: go/vt/mysqlctl/builtinbackupengine.go
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: OK, I did a refactor and created a `TimeoutCloser`, please take a look. The `TimeoutCloser` I created does not take a `context`, because the timeout is only relevant for the `Close()` function, and not for the general lifetime of the closer.

**Comment by shlomi-noach** [comment]
I'm looking into @deepthi's comments.

**Comment by shlomi-noach** [comment]
I tend to agree that `Close()` should not be abandoned, and that an error on `Close()` should be considered a backup/restore failure.

**Comment by shlomi-noach** [comment]
No opinion on the timeout duration. I don't have a good sense of how long it makes sense for an external/3rd party closer to `Close()`. A minute? 5 minutes? 10 minutes?

**Comment by shlomi-noach** [comment]
I've made some changes that make this PR a bit more like https://github.com/vitessio/vitess/pull/5193.
However, looking at https://github.com/vitessio/vitess/pull/5193/files, I want to say I think it also abandons the `Close()` function. I has a timeout just as this PR does, albeit implemented differently. And the only thing that happens if the timeout expires, is that `5193` calls the context's `cancel` function. However, I'm not sure the `cancel` function is at all relevant at the time of call. The only operation it could affect is `addStripeFiles` and in my understanding that function is long since completed.
I'm still looking into this though.

**Comment by maxenglander** [comment]
Hey @shlomi-noach

The way that it is made non-blocking is through `context.Context` propagation:

 1. `context.Context` is passed to `AddFile`. That context is supposed to be attached to the storage handle returned from `AddFile`. For example look at how this is handled in the `gcsbackupstorage`: https://github.com/vitessio/vitess/blob/70a94669826071e31f4ac14047de911753f4af5d/go/vt/mysqlctl/gcsbackupstorage/gcs.go#L96-L103
 2.  Under the hood, the `gcsbackupstorage` is using [this API](https://cloud.google.com/go/docs/reference/cloud.google.com/go/storage/latest#cloud_google_com_go_storage_ObjectHandle_NewWriter), which has this important note:
     > It is the caller's responsibility to call Close when writing is done. To stop writing without saving the data, cancel the context.
 3. In other words, the GCS object `Close()` method is responsive to cancellation of context that was passed to the object during creation.
 4. If `cancelAddFiles()` is called, `closeFile` will end early with an error because the GCS objects `Close()` early with a context cancelled error.

Note that this entire mechanism is dependent on `backupstorage` implementation correctly handling context cancellations, which not all backup storage implementations actually do (:couch: `s3backupstorage`).

**Comment by deepthi** [comment]
> No opinion on the timeout duration. I don't have a good sense of how long it makes sense for an external/3rd party closer to Close(). A minute? 5 minutes? 10 minutes?

I don't know either. I'd say it's worth waiting a minute. 3 seconds is definitely too short. Anything that involves a network call should not timeout in less than 5 seconds. That's a rule of thumb based on TCP timeouts.

File: go/ioutil/timeout_closer.go
Comment (mdlayher) [inline go/ioutil/timeout_closer.go:27]: You don't want to embed the closer here, just use the unexported field.

File: go/ioutil/timeout_closer.go
Comment (mdlayher) [inline go/ioutil/timeout_closer.go:41]: The channel sender goroutine should be the only one to `defer close(done)` or else you may return on context cancel, close this, and then the goroutine may try to send on a closed channel and panic.

File: go/ioutil/timeout_closer.go
Comment (mdlayher) [inline go/ioutil/timeout_closer.go:32]: More generally you could make this accept a context rather than synthesizing one with a timeout, but that's a nit.

File: go/ioutil/timeout_closer_test.go
Comment (mdlayher) [inline go/ioutil/timeout_closer_test.go:49]: `assert.ErrorIs(t, err, context.DeadlineExceeded)`

File: go/ioutil/timeout_closer.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go:27]: fixed

File: go/ioutil/timeout_closer.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: fixed

File: go/ioutil/timeout_closer.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: fixed

File: go/ioutil/timeout_closer_test.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer_test.go]: fixed

**Comment by shlomi-noach** [comment]
>  My recommendation is to rename builtinbackupengine2_test.go to builtinbackupengine_test.go and builtinbackupengine_test.go to backup_blackbox_test.go or something similar.

Done.

**Comment by shlomi-noach** [comment]
@maxenglander thank you - and you are right.

**Comment by shlomi-noach** [comment]
>  The only other thing we need to do is to implement the changes in this PR for xtrabackup as well, because similar code exists there.

Done

**Comment by shlomi-noach** [comment]
Reduced `closeTimeout` to `1min`. This applies both to `XtrabackupEngine` and `BuiltinEngine`.

File: go/vt/mysqlctl/backup_blackbox_test.go
Comment (shlomi-noach) [inline go/vt/mysqlctl/backup_blackbox_test.go:1]: This was renamed from `go/vt/mysqlctl/builtinbackupengine_test.go`

File: go/vt/mysqlctl/builtinbackupengine2_test.go
Comment (shlomi-noach) [inline go/vt/mysqlctl/builtinbackupengine2_test.go:1]: File is renamed to `go/vt/mysqlctl/builtinbackupengine_test.go`

**Comment by shlomi-noach** [comment]
I seem to have broken backup related tests. Looking into.

File: go/ioutil/timeout_closer.go
Comment (deepthi) [reply go/ioutil/timeout_closer.go]: I thought what @mdlayher meant was that you create the context with timeout outside the call to `NewTimeoutCloser` and assume that the passed in context already has a timeout. So instead of `NewTimeoutCloser(closer, timeout)`, you have `NewTimeoutCloser(ctx, closer)`.

**Review by deepthi — APPROVED** [review]
This all looks mostly fine to me. The one thing that isn't very clear/ideal/idiomatic is covered by https://github.com/vitessio/vitess/pull/13668#discussion_r1282208996

Of course tests are telling us something, so we have to make them happy before merging.

File: go/ioutil/timeout_closer.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: The problem I see with that approach is that the timeout should really only apply to the `Close()` function, not to the overall lifespan of the closer object, which is created way before it's being `Close()`d.

Comment (maxenglander) [reply go/ioutil/timeout_closer.go]: I think the fact that this helper is wrapping the `io.Closer` interface makes it sort of clear that this only applies to the `Close` function. We could rename `NewTimeoutCloser` to something more descriptive, like `ClosesWithTimeout`.

**Comment by shlomi-noach** [comment]
Found the reason for the test failures: by _deferring_ the `Close()` function, we inadvertently also changes the logic ordering: the code that was previously non-deferred ran first, now it runs last.

Specifically, this piece of code has to run _after_ the decompressor closes, or else the decompressor receives an unexpected EOF:

```go
	// Close the backupPipe to finish writing on destination.
	if err = bw.Close(); err != nil {
		return vterrors.Wrapf(err, "cannot flush destination: %v", name)
	}

	if err := br.Close(); err != nil {
		return vterrors.Wrap(err, "failed to close the source reader")
	}
```

To fix that I introduced an inner function `createAndCopy` inside `backupFile`. Not the prettiest but it seems to to the job.

File: go/ioutil/timeout_closer.go
Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: I'm open to any change. To be fair, I don't think this functionality justifies a new `type`. Teh entire logic is in one function; maybe we should refactor this back into some `ioutil.CloseWithTimeout()` function. I don't mind either way and I don't have any opinions on naming.

Comment (mdlayher) [reply go/ioutil/timeout_closer.go]: Ah I see now. What I had intended was:

```go
func NewTimeoutCloser(ctx context.Context, closer io.Closer, timeout time.Duration) *TimeoutCloser {
```

Shoudl be:

```go
func NewContextCloser(ctx context.Context, closer io.Closer) *ContextCloser {
```

Which means you don't particularly care if there was a timeout or the user canceled it, the end result is the same. A context cancelation will unblock the operation, and your API can be entirely agnostic of the timeout duration value.

Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: @mdlayher right -- so again the problem I see here, is that the obejct is created at some point in time, let's say `12:00`. Then, some files are compressed/decompressed, saved/loaded. It's now `12:17`. Now we want to `Close()` them. Now say we want a `1min` timeout for the `Close()` function. But `17min` have already passed. If we were to supply `context` into `NewContextCloser()`, then by the time we call `Close()` the context will already have expired `16min` ago, giving us zero time to `Close()`. 

So the issue is we really have to start the counter/timer as we begin the call to `Close()`. Which is why the `TimeoutCloser` has to be aware of the timeout.

To top that, I again think we're better off with just a `CloseWithTimeout()` function, because this entire object grabs so much information upon construction, that it only uses in the `Close()` function.

Comment (shlomi-noach) [reply go/ioutil/timeout_closer.go]: I'm thinking to now just keep it as it is. If anyone feels strongly about how to re-implement this, please feel absolutely free to modify, I don't really mind either way (other than we must ensure the timout only applies to the runtime of `Close()` as described above).

**Comment by shlomi-noach** [comment]
Going to merge this in the next couple hours. I think the core behavior changes are all agreed on, and what remaining discussions around the Closer implementation have converged to "no strong opinions" and we can always iterate on those in a follow up PR. If there's otherwise any outstanding concerns please comment.

**Pull request merged**
